### PR TITLE
GH-72: Add TCP Source

### LINKED ISSUE #72: Port TCP as s-c-s source
As a developer, I'd like to move *tcp* module from XD to s-c-s repo, so I can use it as source to build streaming pipeline.
**Acceptance:**
* *tcp* as source module is available in *spring-cloud-stream-modules* repo 
* maven repo includes *tcp* module with respective group/artifact/version detail

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves spring-cloud/spring-cloud-stream-modules#72

### CODE REVIEW

File: common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/EncoderDecoderFactoryBean.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+public class EncoderDecoderFactoryBean extends AbstractFactoryBean<AbstractByteArraySerializer>
+		implements ApplicationEventPublisherAware {
+
+	private final Encoding encoding;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	public EncoderDecoderFactoryBean(Encoding encoding) {
+		this.encoding = encoding;
```
Comment (artembilan) [inline common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/EncoderDecoderFactoryBean.java:46]: `Assert.notNull()`

File: tcp-source/pom.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	</properties>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.springframework.cloud.stream.module</groupId>
+			<artifactId>spring-cloud-stream-modules-common-configuration</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.integration</groupId>
+			<artifactId>spring-integration-test</artifactId>
```
Comment (artembilan) [inline tcp-source/pom.xml:27]: Is it really that this one isn't included somewhere in parents?

File: tcp-source/src/test/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceTests.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		public void test() throws Exception {
+			assertThat(this.connectionFactory, Matchers.instanceOf(TcpNioServerConnectionFactory.class));
+			assertTrue(TestUtils.getPropertyValue(this.connectionFactory, "lookupHost", Boolean.class));
+			assertTrue(TestUtils.getPropertyValue(this.connectionFactory, "usingDirectBuffers", Boolean.class));
+			assertEquals(123, TestUtils.getPropertyValue(this.connectionFactory, "soTimeout"));
+		}
+
+	}
+
+	@IntegrationTest({ "port = 0" })
```
Comment (artembilan) [inline tcp-source/src/test/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceTests.java]: You can move `port` option to the `TcpSourceTests` level:

```
@WebIntegrationTest("port = 0")
```

and remove it from each these particular `@IntegrationTest`.
See `CassandraSinkIntegrationTests`.

**Comment by artembilan** [comment]
This module isn't included to the parent pom as `module`.

File: tcp-source/pom.xml (hunk 2 above)
Comment (artembilan) [inline tcp-source/pom.xml:27]: So, yes, it is a transitive through the `spring-cloud-stream-modules-common-configuration` - > `spring-cloud-stream-modules-test-support`.
Although I wonder if we really need here such a big dependency tree provided by the `spring-cloud-stream-modules-common-configuration`...
However it doesn't hurt: just only IDE freezing during artifacts resolution :smile:

Comment (artembilan) [inline tcp-source/pom.xml:27]: Sorry, never mind. 
`SI-test` isn't picked up...

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * A source module that receives data over TCP.
+ *
+ * @author Gary Russell
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(TcpSourceProperties.class)
+public class TcpSource {
+
+	@Autowired
+	private Source channels;
```
Comment (artembilan) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: Don't we need here something like:

```
@Bindings(TcpSource.class)
```

?
Although it works well as is!
From other side I would like to see code which takes care about `@Bindings` `@Qualifier` injection...

File: tcp-source/pom.xml (hunk 2 above)
Comment (garyrussell) [inline tcp-source/pom.xml:27]: STS somehow magically added this for me when I used `TestUtils`; I just added the scope.

Comment (artembilan) [inline tcp-source/pom.xml:27]: Yes, it is some magic in the IDE with the Maven plugin. I like it to :smile: 
As I said: never mind and leave it as is.

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java (hunk 4 above)
Comment (garyrussell) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: Hmmm - some have it, some don't @mbogoevici what's the deal? Is `@Bindings` optional?

**Comment by garyrussell** [comment]
Pushed

File: common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/AbstractTcpConnectionFactoryProperties.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ericbottard) [inline common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/AbstractTcpConnectionFactoryProperties.java]: wrong indent on licenses

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * Properties for the TCP Source.
+ *
+ * @author Gary Russell
+ *
+ */
+@ConfigurationProperties
+public class TcpSourceProperties extends AbstractTcpConnectionFactoryProperties {
+
+	/**
+	 * The decoder to use when receiving messages. CRLF, LF, NULL, STXETX, L1, L2, L4, RAW.
```
Comment (ericbottard) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java]: No need to specify the possible values. These are discovered because it's an enum

Comment (garyrussell) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java]: I didn't see that in the json, though.

File: common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/AbstractTcpConnectionFactoryProperties.java (hunk 5 above)
Comment (garyrussell) [inline common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/tcp/AbstractTcpConnectionFactoryProperties.java]: Interesting; this file started life as your `AbstractTcpConnectionFactoryOptionsMetadata` in XD; hence your author credit :smile:

**Comment by artembilan** [comment]
I'm fine with the fix, if that.
Not related to the PR, of course, but I'd like to header about `@Bindings`...

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java (hunk 4 above)
Comment (mbogoevici) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: Normally, it's not needed if there's a single `Source` in the context.

Comment (mbogoevici) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: I prefer putting it, don't mind the extra annotation, and makes a bit clearer what it is.

Comment (artembilan) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: And what it is? 
Some Dave Syer's magic again ? :wink: 
Where to read about this `@Bindings` and so on? Who processes them ?

Thank you, of course, for your feedback!

Comment (mbogoevici) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: No problem, it's part of the reference doc for Spring Cloud Stream (which is very much WIP): https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc

TL;DR `@Bindings` is a qualifier which is automatically attached to the proxies generated by `@EnableBindings`, so it's subject to the normal autowiring rules - if there's only bean of a type, no qualifier is needed.

Comment (artembilan) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: Awesome! 
Old new magic.
Thank you @mbogoevici!

I can merge the PR, BTW, just let me know if you all don't have still any concern there.

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java (hunk 6 above)
Comment (ericbottard) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java]: It's not in the json, but code completion knows how to discover enum values. see https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/EnumValueHintProvider.java

Comment (garyrussell) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java]: I see it is partially expanded in the STS properties editor, but the list is truncated with an elipsis `...`.

Perhaps we need to talk to the STS team. I will remove for now.

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java (hunk 4 above)
Comment (mbogoevici) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSource.java:43]: I don't think that the lack of the annotation is a particular issue.

**Comment by garyrussell** [comment]
Rebased, pushed.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud/spring-cloud-stream-modules/commit/23461fce07cd368f2573e88b4f9c33d1e7f940a8

File: tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java (hunk 6 above)
Comment (garyrussell) [inline tcp-source/src/main/java/org/springframework/cloud/stream/module/tcp/source/TcpSourceProperties.java]: I am talking about the content assist in the IDE, not the CLI. I raised a [JIRA issue against STS](https://issuetracker.springsource.com/browse/STS-4291).
